### PR TITLE
wip: breakTextFragment: performance tuning for large text string

### PR DESCRIPTION
**Pull request by ykuzmenko73**
Current tuning result for breakTextFragment with test string 31217 length:
String 31217 length:
master: 1260 seconds
branch: 0.02 seconds

String 312169 length:
master: newer
branch: 0.22 seconds

comparing performance reports for string 9365 length:
master: 35 sec
image

branch, 312169 length:
image

### CODE REVIEW

**Comment by ykuzmenko73** [comment]
[Tuning breakTextFragment - getCharCatSplitPos / getUnicodeDirection](https://trello.com/c/6oVo5aJF/20686-tuning-breaktextfragment-getcharcatsplitpos-getunicodedirection)

File: lib/text/bidi_text.flow [hunk 1]
```diff
@@ -26,19 +26,80 @@ export {
...
+	);
+
+	makeCharCatSplitPosText(t : string) -> CharCatSplitPosText;
+}
+
+makeCharCatSplitPosText(t : string) -> CharCatSplitPosText {
+	t_space = " " + t;
+	space2_t = "  " + t;
+	space_t = t + " ";
+
```
Comment (yegordovganich) [inline lib/text/bidi_text.flow:44]: 3 lines above ^ - debug code?

File: lib/text/bidi_text.flow [hunk 2]
```diff
@@ -26,19 +26,80 @@ export {
 
 	// Returns a reversed list with elements of original list wrapped into a pair with text flow direction
 	markListToBiDiReverse(list: List<?>, getText: (?) -> string) -> List<Pair<?, int>>;
+	CharCatSplitPosText(
```
Comment (alstrup) [inline lib/text/bidi_text.flow:29]: Add comments to this structure to explain what each field is

File: lib/text/bidi_text.flow [hunk 3]
```diff
@@ -26,19 +26,80 @@ export {
...
+	makeCharCatSplitPosText(t : string) -> CharCatSplitPosText;
+}
+
+makeCharCatSplitPosText(t : string) -> CharCatSplitPosText {
+	t_space = " " + t;
+	space2_t = "  " + t;
+	space_t = t + " ";
+
+	lenText = strlen(t);
+	text = generate(0, lenText, \i -> getCharCodeAt(t, i));
```
Comment (alstrup) [inline lib/text/bidi_text.flow:46]: There is s2a or similar, which might be the same?

File: lib/text/bidi_text.flow [hunk 4]
```diff
@@ -26,19 +26,80 @@ export {
...
+	space = getCharCodeAt(" ", 0);
```
Comment (alstrup) [inline lib/text/bidi_text.flow:47]: Really? We know this is 32, so why bother?

File: lib/text/unicodecharacters.flow [hunk 5]
```diff
@@ -63,10 +65,12 @@ export {
 	// Get the unicode character alphabet by character code.
 	// Returns three-letter string or empty.
 	getUnicodeAlphabet(code : int) -> string;
+	getUnicodeAlphabetEvaluate(code : int) -> string;
```
Comment (alstrup) [inline lib/text/unicodecharacters.flow:68]: Strange names without commnt. Also, do these need to be exported?

**Review by alstrup — APPROVED** [review]
Hard to follow what is going on, but the results speak for themselves. So if it is correct, go for it

File: lib/text/bidi_text.flow (hunk 1 above)
Comment (ykuzmenko73) [reply lib/text/bidi_text.flow]: fixed

File: lib/text/bidi_text.flow (hunk 2 above)
Comment (ykuzmenko73) [reply lib/text/bidi_text.flow:29]: done

File: lib/text/bidi_text.flow (hunk 4 above)
Comment (ykuzmenko73) [reply lib/text/bidi_text.flow:43]: done

File: lib/text/bidi_text.flow (hunk 3 above)
Comment (ykuzmenko73) [reply lib/text/bidi_text.flow:42]: done

File: lib/text/unicodecharacters.flow [hunk 6]
```diff
@@ -185,10 +186,17 @@ ucLookupTable(table, c, fallback) {
 			subtable[lo]
 		else fallback;
 	} else fallback;
-} 
+}
 
+UnicodeDirectionCache = {
```
Comment (alstrup) [inline lib/text/unicodecharacters.flow:191]: This implies a runtime initialization cost for all flow programs that include this. Is that intended? Maybe use onlyOnce?
